Remove Python dependency from `EquivalenceLibrary`

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
The following commits remove the Python dependency from `EquivalenceLibrary` mainly from the following things:
- The `raise_if_param_mismatch` function uses rust native parameters and will filter out from the parameters included by the gate.
- Use `CircuitData` when using Rust native methods.
- Exposes the `new()` method to the rest of the crates.

### Details and comments
~~This is built on top of #14799 and should be rebased once it merges.~~ Rebased!

raynelfss committed:
Initial: Make `add_equivalence` rust native.
The following commits remove the python dependency from the `raise_if_param_mismatch` function from the `EquivalenceLibrary` which removes most of the usage of the `py` token in this structure.

raynelfss committed:
Refactor: Use `CircuitData` for all operations

raynelfss committed:
Chore: Change visibility of `EquivalenceLibrary::new()`.

mtreinish commented on crates/transpiler/src/equivalence.rs:
```diff
@@ -637,14 +635,13 @@ impl EquivalenceLibrary {
         for equiv in entry.iter() {
             raise_if_shape_mismatch(gate, equiv)?;
-            raise_if_param_mismatch(py, params, equiv.0.unsorted_parameters(py)?)?;
+            raise_if_param_mismatch(params, equiv.parameters())?;
```
Hmm, I think this was an oversight in #14799 this is actually going to sort the parameters in the circuit which isn't really necessary for a mismatch check. What we really need is a Rust method on `CircuitData` that returns the iterator from calling `CircuitData.param_table::iter_symbols()`. It should be a 3 line method:

```rust
pub fn iter_parameters(&self) -> impl Iterator<Item = &Symbol> {
    self.param_table.iter_symbols()
}
```

raynelfss replied:
Added in 10cf87a, I do think it falls outside of the scope of the PR, but since it is a smaller addition, I think it can be kept.

mtreinish commented on crates/transpiler/src/equivalence.rs:
```diff
@@ -590,16 +590,15 @@ impl EquivalenceLibrary {
         let key: Key = Key::from_operation(gate);
         let equiv = Equivalence {
-            circuit: equivalent_circuit.clone(),
+            circuit: CircuitFromPython(equivalent_circuit.clone()),
```
Why do you need to use the newtype here? It seems a bit odd that this is needed for anything anymore once we're in a Rust space function?

Like I assume it's because you have `IntoPyObject` implemented for the newtype and you're using that as the return type from a a pymethod. But I think a stricter separation where you just normalize on using `CircuitData` insde the rust domain makes the most sense. You can do the conversion to a Python space `QuantumCircuit` in the python interface layer if you're returning `Equivalence` anywhere directly.

raynelfss replied:
FIxed in 10cf87a

raynelfss committed:
Fix: Use `CircuitData` in `Equivalence`:
- Stop using `CircuitFromPython` for internal storage of `Equivalence`.
- Make the necessary modifications on the rest of the files.

mtreinish commented on crates/transpiler/src/equivalence.rs:
```diff
@@ -709,46 +713,50 @@ impl EquivalenceLibrary {
+        })
+        .collect();
+    let circuit_parameters: HashSet<Symbol> = circuit_parameters.cloned().collect();
+    if parsed_gate_params != circuit_parameters {
```
Do you need a second allocation here, can you just do?:

```suggestion
    if !circuit_parameters.any(|x| parsed_gate_params.contains(x)) {
        let circuit_parameters: HashSet<Symbol> = circuit_parameters.cloned().collect();
```

raynelfss replied:
Yes! I would also have to check for the sizes before doing this but it works!

raynelfss replied:
Corrected in 5cf3db5ddb1b7fff29a21f8ddb462e206f6d7554

mtreinish commented on crates/transpiler/src/equivalence.rs:
```diff
@@ -709,46 +713,50 @@ impl EquivalenceLibrary {
             of different parameters. Gate params: {gate_params:?}. \
-            Circuit params: {circuit_parameters:?}."
+            Circuit params: {:?}.",
+            circuit_parameters,
```
Why did you change this? Modern clippy will complain.

raynelfss replied:
Fixed in 5cf3db5ddb1b7fff29a21f8ddb462e206f6d7554

raynelfss committed:
Fix: Address Matt's comments.

mtreinish commented on crates/transpiler/src/equivalence.rs:
```diff
@@ -149,6 +153,11 @@ impl Equivalence {
+    #[getter]
+    fn get_circuit<'py>(slf: &'py Bound<'py, Self>) -> CircuitFromPython {
+        CircuitFromPython(slf.borrow().circuit.clone())
+    }
```
Why not just?:

```suggestion
    #[getter]
    fn get_circuit(&self) -> CircuitFromPython {
        CircuitFromPython(self.circuit.clone())
    }
```
It's effectively the same thing but with less boilerplate right?

mtreinish reviewed (commented):
I think this is basically ready, just one question/suggestion inline but otherwise we can go ahead and merge this.

raynelfss committed:
Update crates/transpiler/src/equivalence.rs

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>
